### PR TITLE
feat(eslint-plugin): [no-floating-promises] add option 'allowForKnownSafePromises'

### LINKED ISSUE #7008: Enhancement: [no-floating-promises] add an 'allowForKnownSafePromises' option
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-floating-promises

### Description

For years, various community members have asked to add a type allowlist/blocklist or other _"stop complaining about these specific areas"_ for `no-floating-promises`:

* #2640
* #3192 _(ish)_
* #4722
* #4791
* #5231
* #5844
* #6017
* #7006

<sub>_(including clearly duplicate issues to show how this is a relatively more requested option, and to surface the different use cases requested for)_</sub>

Out of those, #5844's `node:test` is most convincing - as `node:test` is stable as of Node 20.0.0 (https://github.com/nodejs/node/issues/46642 -> https://github.com/nodejs/node/pull/46983), and returns `Promise<void>` for tests: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/8730660c37a33dc0768b3f1e36b679294acc4988/types/node/test.d.ts#L134.

I propose we adopt the format discussed in #6017 & implemented in #4436 ([see v6 docs](https://v6--typescript-eslint.netlify.app/rules/prefer-readonly-parameter-types/#allow)) as a new `allow` option for `no-floating-promises`. We'd want to add some very explicitly, strongly worded descriptions in the docs as to why it's not an option we recommend using lightly.

### Fail

```typescript
async function test(name: string, executor: () => Promise<void>) {
  // ...
}

test("example", () => {});
```

### Pass

```typescript
import test from 'node:test';

test("example", () => {});
```

### Additional Info

cc @armano2 @bradzacher as past people who, like me, have previously disagreed with the suggestion of adding this option.

Edit: to be clear, this is roughly what users would be able to specify in their ESLint config _(pending naming bikeshedding)_:

```js
"@typescript-eslint/no-floating-promises": ["error", {
  "allowForKnownSafePromises": [
    { "from": "package", "name": "SafePromise", "package": "some-fancy-package"
  ]
}]
```

### PR DESCRIPTION
**Pull request by kirkwaiblinger**
## PR Checklist

- [x] Addresses an existing open issue: fixes #7008
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

This PR is a continuation of https://github.com/typescript-eslint/typescript-eslint/pull/8502, with a few finishing touches.

A big thank you to @arka1002, who is the primary author of this work! ❤️ 

Co-authored-by: @arka1002

### CODE REVIEW

File: packages/eslint-plugin/docs/rules/no-floating-promises.mdx [hunk 1]
```diff
@@ -119,6 +119,54 @@ await (async function () {
...
+
+This option takes an array of type specifiers to consider safe.
+Each item in the array must have one of the following forms:
+
+- A type defined in a file (`{ from: "file", name: "Foo", path: "src/foo-file.ts" }` with `path` being an optional path relative to the project root directory)
+- A type from the default library (`{ from: "lib", name: "PromiseLike" }`)
+- A type from a package (`{ from: "package", name: "Foo", package: "foo-lib" }`, this also works for types defined in a typings package).
+
+Additionally, a type may be defined just as a simple string, which then matches the type independently of its origin.
+
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/docs/rules/no-floating-promises.mdx:134]: We'll eventually have a consolidated place to describe this shared format... but for now, I really don't want to encourage people to do this.

```suggestion
```

https://github.com/typescript-eslint/typescript-eslint/pull/8502#discussion_r1541768874

File: packages/eslint-plugin/tests/rules/no-floating-promises.test.ts [hunk 2]
```diff
@@ -504,6 +499,209 @@ void promiseArray;
 ['I', 'am', 'just', 'an', 'array'];
       `,
     },
+    // type annotations on variables containing promises
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/tests/rules/no-floating-promises.test.ts:502]: Nit: other rule tests -both in this file and mostly in general for us- don't do this

```suggestion
```

IMO when there are a whole bunch of rule tests like this, it should be apparent what each one / group is testing. They tend to subtly change over time.

...that being said, I am finding some of the more nuanced cases below to be useful. So 🤷 for this I don't feel strongly where the line between "do" and "don't" is.

File: packages/type-utils/tests/TypeOrValueSpecifier.test.ts [hunk 3]
```diff
@@ -202,6 +202,11 @@ describe('TypeOrValueSpecifier', () => {
         'type Foo = {prop: string}; type Test = Foo;',
         { from: 'file', name: 'Foo', path: 'tests/fixtures/file.ts' },
       ],
+      [
+        'type Foo = Promise<number> & {hey?: string}; let foo: Foo = Promise.resolve(5); type Test = typeof foo;',
+        // type.getSymbol() doesn't work here, .aliasSymbol does
```
Comment (JoshuaKGoldberg) [inline packages/type-utils/tests/TypeOrValueSpecifier.test.ts:207]: Nit: this feels pretty disconnected / out-of-context to place in the test. If there's a deeper technical explanation that's helpful to mention, I'd say it'd belong in the implementation.

```suggestion
```

**Review by JoshuaKGoldberg — DISMISSED** [review]
🔥 🔥 🔥 

So pumped to see this land! I like how it ended up being a mostly additive change. This is great. Nice job @arka1002 + @kirkwaiblinger! 

Nothing blocking from me - I'm happy to see this merged asap.

Skeleton putting on a pair of sunglasses on top of an oily moving background. Caption: "hell yeah"

File: packages/eslint-plugin/docs/rules/no-floating-promises.mdx (hunk 1 above)
Comment (kirkwaiblinger) [reply packages/eslint-plugin/docs/rules/no-floating-promises.mdx:134]: ah, sorry, missed that comment. Will revert.

File: packages/eslint-plugin/tests/rules/no-floating-promises.test.ts (hunk 2 above)
Comment (kirkwaiblinger) [reply packages/eslint-plugin/tests/rules/no-floating-promises.test.ts]: trimmed em down a bit 👍

File: packages/type-utils/tests/TypeOrValueSpecifier.test.ts (hunk 3 above)
Comment (kirkwaiblinger) [reply packages/type-utils/tests/TypeOrValueSpecifier.test.ts]: agreed, removed.

**Review by JoshuaKGoldberg — APPROVED** [review]
✅ 🤩 ❗❗❗ 

So excited to have this merged! Thanks!!

Marking as `1 approval`, with a calendar reminder on myself to merge Monday morning if no blocking feedback has been left by then!

**Review by arka1002 — APPROVED** [review]

**Pull request merged**